### PR TITLE
[Admin] DataObject Grid - filter by relation fields

### PR DESCRIPTION
**Pull request by fszenborn**
## Additional info  
Could someone review/test my changes and leave some feedback or suggestions?
Here is the quick demo of the functionality:
https://user-images.githubusercontent.com/54931392/123623624-f6b28080-d80d-11eb-8274-8c091e409235.mp4
Supported field types are:
one to many
many to many
many to many object
advanced many to many
advanced many to many object
image
advanced image

### CODE REVIEW

**Comment by dvesh3** [comment]
@fszenborn I just had a look at the demo video and the feature looks great 🤩 we'll review it soon. Meanwhile could you please check why tests are failing? thanks! great work 👏  

One more thing, please rebase it on 10.x, since this a new feature.

**Comment by dpfaffenbauer** [comment]
THIS IS 🔥

**Comment by dvesh3** [comment]
@fszenborn please rebase it on branch 10.x, so that we can start the review. thanks!

**Comment by fszenborn** [comment]
Sorry for delay - I will do it soon. I was on vacation.

**Comment by dvesh3** [comment]
> Sorry for delay - I will do it soon. I was on vacation.

@fszenborn hope you had a great one 😉 no worries, take your time!

File: bundles/AdminBundle/Resources/public/js/pimcore/element/helpers/gridColumnConfig.js [hunk 1]
```diff
@@ -362,6 +380,95 @@ pimcore.element.helpers.gridColumnConfig = {
...
+        }
+
+        var fieldInfo = this.grid.getColumns()[columnIndex].config;
+
+        if((this.objecttype === "object") || (this.objecttype === "variant")) {
+            if (!fieldInfo.layout || !fieldInfo.layout.layout) {
+                return;
+            }
+
+            if (fieldInfo.layout.layout.noteditable) {
```
Comment (kubaplas) [inline bundles/AdminBundle/Resources/public/js/pimcore/element/helpers/gridColumnConfig.js]: @fszenborn this piece of code causes filters to not work on non-editable fields. I guess you didn't remove it when copying code from batch edit.

File: models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        if ($operator == '=') {
+            $value = "'%," . $value . ",%'";
+            return '`' . $name . '` LIKE ' . $value . ' ';
+        } elseif ($operator == 'LIKE') {
+            $result = $name . " IS NULL";
+            $values = explode(',', (string)$value ?? '');
+            if (is_array($values) && !empty($values)) {
+                $fieldConditions = [];
+                foreach ($values as $value) {
```
Comment (jdevries14) [inline models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php]:
```
$fieldConditions = array_map(function ($value) use ($name) {
    return '`' . $name . "` LIKE '%" . $value . "%' ";
}, array_filter($values));
```

File: models/DataObject/ClassDefinition/Data/ManyToManyRelation.php [hunk 3]
```diff
@@ -942,4 +942,17 @@ public function addListingFilter(DataObject\Listing $listing, $data, $operator =
...
     }
+
+    /**
+     * Filter by relation feature
+     * @param array|string $value
+     * @param string       $operator
+     * @param array        $params
+     * @return string
+     */
+    public function getFilterConditionExt($value, $operator, $params = [])
```
Comment (jdevries14) [inline models/DataObject/ClassDefinition/Data/ManyToManyRelation.php:976]: seeing the same code in lots of classes, why not use a trait?

File: bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js [hunk 4]
```diff
@@ -100,6 +101,17 @@ pimcore.object.tags.manyToManyRelation = Class.create(pimcore.object.tags.abstra
         };
     },
 
+    getRelationFilterCondition: function (editor) {
+        var filterResult = [];
+            editor.store.getData().items.forEach(
```
Comment (jdevries14) [inline bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js]: return ....map();

**Comment by brusch** [comment]
@fszenborn please check feedback of @jdevries14 Thanks!

File: models/DataObject/ClassDefinition/Data/ManyToManyRelation.php (hunk 3 above)
Comment (fszenborn) [reply models/DataObject/ClassDefinition/Data/ManyToManyRelation.php:976]: I have used trait call for the same piece of code but it differs a bit because of database implementation of various relation types that is why I left it as is

File: models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php (hunk 2 above)
Comment (fszenborn) [reply models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php]: Thanks, just changed to map

File: bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js (hunk 4 above)
Comment (fszenborn) [reply bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js]: changed to use map()

**Comment by fszenborn** [comment]
@brusch @dvesh3 I have just rebased on 10.x and applied suggestions from @jdevries14

**Comment by dvesh3** [comment]
> @brusch @dvesh3 I have just rebased on 10.x and applied suggestions from @jdevries14

@fszenborn thanks for your changes! After changing the base to 10.x, there's a conflict. Could you please resolve it?

File: models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    public function getRelationFilterCondition($value, $operator, $name)
+    {
+        if ($operator == '=') {
+            return '`' . $name . '` LIKE ' . "'%," . $value . ",%'";
+        }
+        $result = $name . " IS NULL";
+        $values = explode(',', (string)$value ?? '');
+        if (is_array($values) && !empty($values)) {
+            $fieldConditions = array_map(function ($value) use ($name) {
+                return '`' . $name . "` LIKE '%" . $value . "%' ";
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php]: This returns wrong dataset, specially in case of ManyToOne relation. so when i select `manufacturer_id` = "4" it returns rows with ids 4,45,46 and so on...
image

Also, for ManyToManyRelation, the condition should be `(categories LIKE '%,558,%'  AND categories LIKE '%,559,%' )`

File: bundles/AdminBundle/Resources/public/js/pimcore/element/helpers/gridColumnConfig.js [hunk 6]
```diff
@@ -362,6 +380,90 @@ pimcore.element.helpers.gridColumnConfig = {
...
+                    }.bind(this)
+                },
+                {
+                    text: t("apply_filter"),
+                    iconCls: "pimcore_icon_filter pimcore_icon_overlay_add",
+                    handler: function () {
+                        if (formPanel.isValid() && typeof fieldInfo.getRelationFilterCondition === "function") {
+                            this.grid.filters.getStore().addFilter(
+                                new Ext.util.Filter({
+                                    operator: "like",
```
Comment (dvesh3) [inline bundles/AdminBundle/Resources/public/js/pimcore/element/helpers/gridColumnConfig.js]: operator is always "like" ..ManyToOne should have '=' as operator.

File: models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+     * Parses filter value of a relation field and creates the filter condition
+     * @param string $value
+     * @param string $operator
+     * @param string $name
+     * @return string
+     */
+    public function getRelationFilterCondition($value, $operator, $name)
+    {
+        if ($operator == '=') {
+            return '`' . $name . '` LIKE ' . "'%," . $value . ",%'";
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php]: we should consider type as well for relation filters

Imagine a case where a relation field has both document & object relations having same id then result set will wrong in case of filter based on just one type.

**Comment by dvesh3** [comment]
@fszenborn please see my comments above. thanks!

File: models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php (hunk 5 above)
Comment (fszenborn) [reply models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php:33]: I will check it out. I must have messed sth up while appling suggestions from jdevries14 review

File: bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js [hunk 8]
```diff
@@ -100,6 +101,15 @@ pimcore.object.tags.manyToManyRelation = Class.create(pimcore.object.tags.abstra
         };
     },
 
+    getRelationFilterCondition: function (editor) {
```
Comment (fszenborn) [inline bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js]: @dvesh3 here is the place where object/asset/document id is bundled with it's type as a reply to https://github.com/pimcore/pimcore/pull/10022#discussion_r719203448

File: models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php (hunk 7 above)
Comment (fszenborn) [reply models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php]: please check this files for the piece that considers type as well: 
bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js
bundles/AdminBundle/Resources/public/js/pimcore/object/tags/advancedManyToManyRelation.js

File: bundles/AdminBundle/Resources/public/js/pimcore/element/helpers/gridColumnConfig.js (hunk 6 above)
Comment (fszenborn) [reply bundles/AdminBundle/Resources/public/js/pimcore/element/helpers/gridColumnConfig.js]: OK, I will change this

File: models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php (hunk 7 above)
Comment (fszenborn) [reply models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php]: for manyToMany and advancedManyToMany it is considered
but for ManyToOne it isn't, you're right.

**Comment by dvesh3** [comment]
@fszenborn any chance to work on the reported issues soon? we should get this done in 10.2 release 🚀  thanks!

File: models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php (hunk 7 above)
Comment (fszenborn) [reply models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php]: @dvesh3 I have just fixed and types are considered for ManyToOne relation right now

File: bundles/AdminBundle/Resources/public/js/pimcore/element/helpers/gridColumnConfig.js (hunk 6 above)
Comment (fszenborn) [reply bundles/AdminBundle/Resources/public/js/pimcore/element/helpers/gridColumnConfig.js]: fixed

File: models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php (hunk 5 above)
Comment (fszenborn) [reply models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php]: @dvesh3 fixed

**Comment by fszenborn** [comment]
@dvesh3 just fixed reported issues

File: bundles/AdminBundle/Controller/Admin/ElementController.php [hunk 9]
```diff
@@ -901,7 +901,7 @@ protected function getNicePathFormatterFieldDefinition($source, $context)
      *
      * @throws \Exception
      */
-    protected function convertResultWithPathFormatter(DataObject\Concrete $source, $context, $result, $targets): array
+    protected function convertResultWithPathFormatter(Element\AbstractElement $source, $context, $result, $targets): array
```
Comment (dvesh3) [inline bundles/AdminBundle/Controller/Admin/ElementController.php]: why this change?

File: models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
+<?php
+/**
+ * @category    pimcore
+ * @date        25/06/2021
+ * @author      Filip Szenborn <fszenborn@divante.co>
+ */
+declare(strict_types=1);
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Data/Extension/RelationFilterConditionParser.php]: please follow coding standards defined here https://github.com/pimcore/pimcore/blob/10.x/.php-cs-fixer.dist.php

Author information will be stored with the commit
```suggestion
declare(strict_types=1);

/**
 * Pimcore
 *
 * This source file is available under two different licenses:
 * - GNU General Public License version 3 (GPLv3)
 * - Pimcore Commercial License (PCL)
 * Full copyright and license information is available in
 * LICENSE.md which is distributed with this source code.
 *
 *  @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
 *  @license    http://www.pimcore.org/license     GPLv3 and PCL
 */
```

File: bundles/AdminBundle/Controller/Admin/ElementController.php (hunk 9 above)
Comment (fszenborn) [reply bundles/AdminBundle/Controller/Admin/ElementController.php]: @dvesh3 Could you suggest the proper way of fixing this? It could be fixed on ExtJS side as well

I did this because on pimcore X I got the folowing error while selecting the elements to be filtered by (caused by nicePathFormatter):
`
Pimcore\Bundle\AdminBundle\Controller\Admin\ElementController::convertResultWithPathFormatter(): 
Argument #1 ($source) must be of type Pimcore\Model\DataObject\Concrete, 
Pimcore\Model\DataObject\Folder given, 
called in /--/pimcore/vendor/pimcore/pimcore/bundles/AdminBundle/Controller/Admin/ElementController.php on line 567`

The error is caused because the editLayout for relation fields, which I am reusing for filter window, sends the source/context information about the "parent" in extJS (usually on the object detail view the parent is the object view). but On the grid the parent is the Folder - not an Concrete instance and that causes this error. 
I haven't got this error on pimcore 6.9 where I actually developed this functionality.

Comment (dvesh3) [reply bundles/AdminBundle/Controller/Admin/ElementController.php]: I would simply return here when the container type is `filterByRelationWindow` (like `batch` container type) https://github.com/pimcore/pimcore/blob/10.x/bundles/AdminBundle/Resources/public/js/pimcore/helpers.js#L2541

We don't need to request nice path for filtering purpose.

Comment (dvesh3) [reply bundles/AdminBundle/Controller/Admin/ElementController.php]: @fszenborn

Comment (fszenborn) [reply bundles/AdminBundle/Controller/Admin/ElementController.php]: @dvesh3 fixed

**Review by dvesh3 — APPROVED** [review]

**Comment by dvesh3** [comment]
@fszenborn great work 👏 thanks a lot!

@kubaplas @jdevries14 thanks for the review 😊

File: bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyObjectRelation.js [hunk 11]
```diff
@@ -88,10 +88,31 @@ pimcore.object.tags.manyToManyObjectRelation = Class.create(pimcore.object.tags.
...
+        if (!filterValues || !Array.isArray(filterValues) || !filterValues.length) {
+            filterValues = null;
+        } else {
+            filterValues = filterValues.map(function (item) {
+                return item.data.id;
+            }).join(',');
+        }
+
+        return new Ext.util.Filter({
+            operator: "like",
```
Comment (cancan101) [inline bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyObjectRelation.js:107]: Why is the type here `string` rather than `int` (or `numeric` or `list`)? (and `like` vs `=`)?
